### PR TITLE
apps/ca: allow CRL lastUpdate/nextUpdate fields to be specified

### PR DESCRIPTION
**Pull request by chrisnovakovic**
When generating a CRL using the `ca` utility, allow values for the `lastUpdate` and `nextUpdate` fields to be specified using the command line options `-crl_lastupdate` and `-crl_nextupdate` respectively.

##### Checklist
- [x] documentation is added or updated

### CODE REVIEW

**Comment by chrisnovakovic** [comment]
I've just emailed in my completed copy of the ICLA.

File: doc/man1/openssl-ca.pod.in
Comment (beldmit) [inline doc/man1/openssl-ca.pod.in:357]: Maybe it's worth replacing the duplicated description of the format by the reference to the previous option?

**Review by beldmit — APPROVED** [review]
LGTM if CLA is in place

File: doc/man1/openssl-ca.pod.in
Comment (chrisnovakovic) [reply doc/man1/openssl-ca.pod.in:354]: I followed the same style as `startdate` and `enddate` for consistency, but I agree that it looks too repetitive - I've simplified the documentation for the two new options.

**Comment by chrisnovakovic** [comment]
Closing and reopening to get @openssl-machine to accept my ICLA.

File: apps/ca.c
Comment (slontis) [inline apps/ca.c:1180]: NIT: indent the params to align with the start of the params above..

File: apps/lib/apps.c
Comment (slontis) [inline apps/lib/apps.c:2710]: days, hours, secs were long params before.. Should they still be?

**Review by slontis — COMMENTED** [review]
Some form of test to at least 'cover' this would be good.
See test/recipes/*.t files.

Adding to an existing .t file is fine..

I assume you have manually verified it works?

File: apps/ca.c
Comment (chrisnovakovic) [reply apps/ca.c]: Fixed in the newest revision.

File: apps/lib/apps.c
Comment (chrisnovakovic) [reply apps/lib/apps.c]: Whoops, my mistake - fixed in the latest revision, thanks.

**Comment by chrisnovakovic** [comment]
Thanks for the review, Shane.

> Some form of test to at least 'cover' this would be good.
> See test/recipes/*.t files.
> 
> Adding to an existing .t file is fine..
> 
> I assume you have manually verified it works?

I'd manually tested all of the code paths, but didn't realise there were tests for the command line interface too. I've added relevant tests to `test/recipes/80-test_ca.t`. While writing those tests, I noticed that `openssl ca` returns 0 if `-crl_lastupdate` or `-crl_nextupdate` are invalid (as it does if `-startdate` or `-enddate` are invalid), but in retrospect I think it makes more sense to return 1 in those cases (it already returns 1 in other failure cases, so returning 0 here seems like a bug). I've made that change in the latest revision, and I can do the same when invalid versions of `-startdate` and `-enddate` are given in a separate PR if you like.

File: test/recipes/80-test_ca.t
Comment (slontis) [inline test/recipes/80-test_ca.t:75]: I probably would have just added a simple extra parameter here that is the expected value 1 or 0 instead of doing the regex..

Comment (chrisnovakovic) [reply test/recipes/80-test_ca.t]: Good point, I probably overthought that a bit. I've replaced that with a single `should_succeed` key.

**Comment by chrisnovakovic** [comment]
I'm experiencing some [odd CI failures on macOS](https://travis-ci.com/github/openssl/openssl/jobs/382550133) that I can't test fixes for locally because I don't have access to a Mac. I'm hoping some opportunistic use of `diag()` will throw some light on what's happening, but given the queue size on AppVeyor right now, this might take another few days to resolve.

**Comment by slontis** [comment]
> I'm experiencing some [odd CI failures on macOS](https://travis-ci.com/github/openssl/openssl/jobs/382550133) t

Your perl skills are better than mine (which is not too hard).
I would be looking to see if you are doing anything different from other .t files.

**Comment by slontis** [comment]
Also using fixup commits via
```
> git commit --fixup  commit_id
> git push
```
Makes it easier to review any changes.. For larger changes this can save quite a bit of time.
(Use >git log to view commit_ids)

When we do the final merge we can squash these automatically..

File: test/recipes/80-test_ca.t
Comment (slontis) [inline test/recipes/80-test_ca.t:283]: I cant see anywhere else that accesses test-runs like this for example

Comment (chrisnovakovic) [reply test/recipes/80-test_ca.t]: Yeah, I think I was trying to write files to a path that doesn't exist in every CI environment by doing that. The latest commit should fix that by not caring too much about where temporary CSRs/certificates/CRLs are written (no other test in this file appears to, so hopefully that's not a problem).

**Review by t8m — APPROVED** [review]
LGTM

File: test/recipes/80-test_ca.t
Comment (beldmit) [inline test/recipes/80-test_ca.t:104]: Is 2009 instead of 2099 intentional?

**Review by beldmit — APPROVED** [review]

File: test/recipes/80-test_ca.t
Comment (chrisnovakovic) [reply test/recipes/80-test_ca.t:104]: Yes - those are two-digit-year timestamps (e.g. 2020-09-01 12:34:56), so OpenSSL should parse them as UTCTimes. The others are four-digit-year, post-2049 timestamps (e.g. 2099-09-01 12:34:56), so OpenSSL should (if I understood correctly) parse them as GeneralizedTimes. Does that make sense?

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by chrisnovakovic** [comment]
Thanks! Do I still need to wait for a review from @slontis, or am I okay to merge this now? (I don't see the merge button, presumably because there's one outstanding review.)

**Comment by beldmit** [comment]
You've got the necessary 2 approves, but the merge is done not by you but by someone from OpenSSL Team members (presumably @t8m or me).

**Comment by beldmit** [comment]
Merged. Many thanks!

**Pull request closed without merging**
